### PR TITLE
Update PUT /filter-outputs/{filter_output_id} Endpoint

### PR DESCRIPTION
**Pull request by j-s-rawat**
### What
https://trello.com/c/a2Hf0SPa/5626-update-put-filter-outputs-filteroutputid-endpoint

1. Filter outputs data model is updated and this forced the changes to be done to the endpoint
2. Updated logic to meet the swagger spec changes 
3. Introduced error handler for Forbidden status

### How to review
The logic should match swagger specifications

### Who can review

Anyone

### CODE REVIEW

File: api/contract.go [hunk 1]
```diff
@@ -68,18 +68,6 @@ type putFilterResponse struct {
 	model.PutFilter
 }
 
-// createFilterOutputResponse is the response body for POST /filters-output
-type createFilterOutputResponse struct {
```
Comment (j-s-rawat) [inline api/contract.go:72]: Removed these as there is no JSON  response for filter output. Just status code is returned

File: datastore/mongodb/filter_outputs.go [hunk 2]
```diff
@@ -30,8 +31,50 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
...
-	if _, err := c.conn.Collection(col.name).UpsertById(ctx, f.ID, bson.M{"$set": f}); err != nil {
-		return errors.Wrap(err, "failed to upsert filter")
+	var docs []model.FilterOutput
+	sc := bson.M{"id": f.ID}
+
+	var err error
+	var num int
+
+	if num, err = c.conn.Collection(col.name).Find(ctx, sc, &docs, mongodb.Limit(1)); err != nil {
+		return errors.Wrap(err, "failed to update filter outputs")
```
Comment (ChrysmOre) [inline datastore/mongodb/filter_outputs.go:41]: Should be `output` (singular)

File: datastore/mongodb/filter_outputs.go [hunk 3]
```diff
@@ -30,8 +31,50 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
...
+		err := &er{
+			err: errors.Errorf("failed to find filter output"),
+		}
+		err.notFound = true
+		return err
+	}
+
+	//a record with stat 'completed' can't be updated further
+	if docs[0].State == model.COMPLETED {
+		err := &er{
```
Comment (ChrysmOre) [inline datastore/mongodb/filter_outputs.go:53]: No need to create the error before returning it here
```suggestion
		return &er{
		  err: errors.Errorf(`filter output is already in "completed" state`),
		  forbidden: true,
		}
		
```

File: datastore/mongodb/filter_outputs.go [hunk 4]
```diff
@@ -30,8 +31,50 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
...
+
+	if rec, err = c.conn.Collection(col.name).Update(ctx, sc, updates); err != nil {
+		return errors.Wrap(err, "failed to update filter outputs")
+	}
+
+	//This should not happen unless the recored is being removed btween the first find and the update.
+	//Check if the update failed sue to search condition.
+	//Update returns no error but the result's MatchedCount has to be checked for number of updated records
+	if rec.MatchedCount < 1 {
+		println("Record not found... Searched ID: ", "`", f.ID, "`")
```
Comment (ChrysmOre) [inline datastore/mongodb/filter_outputs.go:72]: Debug?

File: datastore/mongodb/filter_outputs.go [hunk 5]
```diff
@@ -30,8 +31,50 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
...
+		err := &er{
```
Comment (ChrysmOre) [inline datastore/mongodb/filter_outputs.go:73]: Again can just return the error straight away here

File: api/filters.go [hunk 6]
```diff
@@ -153,8 +153,8 @@ func (api *API) submitFilter(w http.ResponseWriter, r *http.Request) {
 	// Just chosen a reasonable looking one.
 	// TODO: RAISE STATE in a round table.
 	filterOutput := &model.FilterOutput{
-		ID:    filter.Dataset.ID,
-		State: "submitted",
+		FilterID: filter.Dataset.ID,
+		State:    "submitted",
```
Comment (ChrysmOre) [inline api/filters.go:157]: You're using a const for completed state, you should probably be consistent and use for others too

File: model/filter.go [hunk 7]
```diff
@@ -8,6 +8,10 @@ import (
 	"go.mongodb.org/mongo-driver/bson/primitive"
 )
 
+const (
+	COMPLETED = "completed"
```
Comment (ChrysmOre) [inline model/filter.go:12]: It's generally a Go convention not to use ALL_CAPS for any variable name including constants (generally standard CamelCase is used).

File: datastore/mongodb/filter_outputs.go (hunk 4 above)
Comment (j-s-rawat) [reply datastore/mongodb/filter_outputs.go:72]: removed

File: features/put_filter_outputs.feature [hunk 8]
```diff
@@ -87,7 +91,7 @@ Feature: Put Filter Outputs Private Endpoints Not Enabled
     ]
     """
 
-  Scenario: PUT filter outputs successfully with existing ID in DB
+Scenario: PUT filter outputs successfully with existing ID in DB
```
Comment (saminahbab) [inline features/put_filter_outputs.feature:94]: The tests look good, perhaps it would be good to involve a test in the event that mongo is down?

Comment (j-s-rawat) [reply features/put_filter_outputs.feature:94]: There is scenario below "Put a filter output with broken mongo db" that covers this

**Review by ChrysmOre — APPROVED** [review]
LGTM

**Pull request merged**